src/workspace.rs

```diff
@@ -0,0 +1,23 @@
+use super::error::{MainError, PnError};
+use lets_find_up::*;
+use std::path::{Path, PathBuf};
+
+const WORKSPACE_MANIFEST_FILENAME: &str = "pnpm-workspace.yaml";
+
+pub fn find_workspace_root(cwd: &dyn AsRef<Path>) -> Result<PathBuf, MainError> {
+    let workspace_manifest_location = find_up_with(
+        WORKSPACE_MANIFEST_FILENAME,
+        FindUpOptions {
+            kind: FindUpKind::File,
+            cwd: cwd.as_ref(),
+        },
+    )
+    .map_err(MainError::from_dyn)?;
+    match workspace_manifest_location {
+        Some(path) => Ok(path
+            .parent()
+            .ok_or(MainError::Pn(PnError::NotInWorkspace))?
+            .to_path_buf()),
+        None => Err(MainError::Pn(PnError::NotInWorkspace)),
+    }
```

Comment:
Alternatively, you can merge two `None`s into one so that there is only one call to `PnError::NotInWorkspace`:

```suggestion
    match workspace_manifest_location.and_then(|x| x.parent()) {
        Some(workspace) => Ok(workspace.to_path_buf()),
        None => Err(MainError::Pn(PnError::NotInWorkspace)),
    }
```

alternate syntax:


```suggestion
    workspace_manifest_location
        .and_then(|x| x.parent())
        .map(|x| x.to_path_buf())
        .ok_or(MainError::Pn(PnError::NotInWorkspace))
```